services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java

```diff
@@ -217,6 +217,79 @@ private void patchGoalstateForNeighbor(NetworkConfiguration networkConfig, Unica
         }
     }
 
+    private void patchGoalstateForRouterSubnet(NetworkConfiguration networkConfig, Map<String, List<String>> hostsSubnets, MulticastGoalState multicastGoalState) throws CacheException {
+        for (Map.Entry<String, List<String>> entry: hostsSubnets.entrySet()) {
+            for (String subnetId : entry.getValue()) {
+                InternalSubnetPorts subnetEntity = subnetPortsCache.getSubnetPorts(subnetId);
+                if (subnetEntity != null) {
+                    if (multicastGoalState.getGoalStateBuilder().getSubnetStatesList().stream()
+                            .filter(e -> e.getConfiguration().getId().equals(subnetEntity.getSubnetId()))
+                            .findFirst().orElse(null) == null) {
+                        Subnet.SubnetConfiguration.Builder subnetConfigBuilder = Subnet.SubnetConfiguration.newBuilder();
+                        subnetConfigBuilder.setRevisionNumber(FORMAT_REVISION_NUMBER);
+                        subnetConfigBuilder.setId(subnetEntity.getSubnetId());
+                        subnetConfigBuilder.setVpcId(subnetEntity.getVpcId());
+                        subnetConfigBuilder.setName(subnetEntity.getName());
+                        subnetConfigBuilder.setCidr(subnetEntity.getCidr());
+                        subnetConfigBuilder.setTunnelId(subnetEntity.getTunnelId());
```

Comment:
If we use builder in this way, every time we add a new field, we will have to come here and add it accordingly. Could we abstract the conversion logics in a single method, then we only need to change the implementation inside the conversion method? @cj-chung

Reply:
abstract the code in this section to `constructSubnetState`.